lens-labels/src/Lens/Labels.hs

```diff
@@ -82,13 +82,26 @@ instance
 #endif
 
 -- | A type class for lens fields.
-class HasLens' s (x :: Symbol) a | s x -> a where
+--
+-- Note: in order to support better type error messages, this class does not have
+-- a fundep relationship to the parameter @a@.  Instead, instances are expected to
+-- be defined such that @a@ is derivable from @s@ and @x@.  For example, to define a
+-- field named @"x"@ of type @Int32@ for the containing type @Foo@:
+--
+-- > instance a ~ Int32 => HasLens' Foo "x" a where ...
+--
+class HasLens' s (x :: Symbol) a where
     lensOf' :: Functor f => Proxy# x -> (a -> f a) -> s -> f s
 
+instance {-# OVERLAPPABLE #-} TypeError (MissingInstanceError s x) => HasLens' s x a where
+    lensOf' = error "Missing HasLens' instance"
+
+type MissingInstanceError s (x :: Symbol) =
+    'Text "Type " ':<>: 'ShowType s ':<>: 'Text " has no field named " ':<>: 'ShowType x
 type ASetter s t a b = LensLike Identity s t a b
 
 (.~), set :: ASetter s t a b -> b -> s -> t
-f .~ x = f %~ const x
+f .~ x' = f %~ const x'
```

Comment:
Why is this prime required? Did we magic up an `x` in global scope somehow?

Reply:
Removed.  (It was part of debugging that I forgot to remove.)